### ISSUE TITLE
aws s3 memory usage high with copy (cross account) 

### ISSUE THREAD

**Issue by bkruger99** [rclone/rclone#2694]
Related to #2691:

May be vendor sdk issue here: https://github.com/aws/aws-sdk-go/issues/2036

Running this (realizing it's not optimal for s3, but need to fix object perms in the process until it's all fixed.): 
`rclone --rc --checkers 400 --transfers 400 copy s3-usw2-bofc:<bucket1> s3-usw2-bofc2:<bucket2>`

Memory usage continues to accumulate.  Here's a snapshot: 
```
root@5zqck:~/pprof# go tool pprof -text http://localhost:5572/debug/pprof/heap
Fetching profile over HTTP from http://localhost:5572/debug/pprof/heap
Saved profile in /home/bkruger/pprof/pprof.rclone.alloc_objects.alloc_space.inuse_objects.inuse_space.012.pb.gz
File: rclone
Type: inuse_space
Time: Oct 24, 2018 at 6:27pm (UTC)
Showing nodes accounting for 11361.14MB, 98.65% of 11517.03MB total
Dropped 173 nodes (cum <= 57.59MB)
      flat  flat%   sum%        cum   cum%
 7230.33MB 62.78% 62.78%  7230.33MB 62.78%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/service/s3/s3manager.(*uploader).init.func1
 3596.77MB 31.23% 94.01%  3596.77MB 31.23%  github.com/ncw/rclone/fs/asyncreader.newBuffer (inline)
  180.52MB  1.57% 95.58%   180.52MB  1.57%  github.com/ncw/rclone/backend/s3.(*Fs).newObjectWithInfo
  164.51MB  1.43% 97.01%   169.04MB  1.47%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/private/protocol/xml/xmlutil.XMLToStruct
  123.01MB  1.07% 98.07%   129.51MB  1.12%  net/textproto.(*Reader).ReadMIMEHeader
   65.51MB  0.57% 98.64%    66.01MB  0.57%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/private/protocol/rest.unmarshalHeaderMap
    0.50MB 0.0043% 98.65%   132.07MB  1.15%  net/http.(*persistConn).readLoop
         0     0% 98.65%   350.56MB  3.04%  github.com/ncw/rclone/backend/s3.(*Fs).List
         0     0% 98.65%   181.02MB  1.57%  github.com/ncw/rclone/backend/s3.(*Fs).itemToDirEntry
         0     0% 98.65%   350.56MB  3.04%  github.com/ncw/rclone/backend/s3.(*Fs).list
         0     0% 98.65%   169.54MB  1.47%  github.com/ncw/rclone/backend/s3.(*Fs).list.func1
         0     0% 98.65%   350.56MB  3.04%  github.com/ncw/rclone/backend/s3.(*Fs).listDir
         0     0% 98.65%   181.02MB  1.57%  github.com/ncw/rclone/backend/s3.(*Fs).listDir.func1
         0     0% 98.65%    80.54MB   0.7%  github.com/ncw/rclone/backend/s3.(*Object).ModTime
         0     0% 98.65%  7233.33MB 62.81%  github.com/ncw/rclone/backend/s3.(*Object).Update
         0     0% 98.65%  7231.33MB 62.79%  github.com/ncw/rclone/backend/s3.(*Object).Update.func2
         0     0% 98.65%    82.54MB  0.72%  github.com/ncw/rclone/backend/s3.(*Object).readMetaData
         0     0% 98.65%    82.54MB  0.72%  github.com/ncw/rclone/backend/s3.(*Object).readMetaData.func1
         0     0% 98.65%  3596.77MB 31.23%  github.com/ncw/rclone/fs/asyncreader.(*AsyncReader).getBuffer
         0     0% 98.65%  3630.81MB 31.53%  github.com/ncw/rclone/fs/asyncreader.(*AsyncReader).init.func1
         0     0% 98.65%  3596.77MB 31.23%  github.com/ncw/rclone/fs/asyncreader.glob..func1
         0     0% 98.65%   350.56MB  3.04%  github.com/ncw/rclone/fs/list.DirSorted
         0     0% 98.65%   349.54MB  3.03%  github.com/ncw/rclone/fs/march.(*March).makeListDir.func1
         0     0% 98.65%   179.01MB  1.55%  github.com/ncw/rclone/fs/march.(*March).processJob.func1
         0     0% 98.65%   170.01MB  1.48%  github.com/ncw/rclone/fs/march.(*March).processJob.func2
         0     0% 98.65%  7234.33MB 62.81%  github.com/ncw/rclone/fs/operations.Copy
         0     0% 98.65%    80.54MB   0.7%  github.com/ncw/rclone/fs/operations.Equal
         0     0% 98.65%    80.54MB   0.7%  github.com/ncw/rclone/fs/operations.NeedTransfer
         0     0% 98.65%    80.54MB   0.7%  github.com/ncw/rclone/fs/operations.equal
         0     0% 98.65%    80.54MB   0.7%  github.com/ncw/rclone/fs/sync.(*syncCopyMove).pairChecker
         0     0% 98.65%  7234.33MB 62.81%  github.com/ncw/rclone/fs/sync.(*syncCopyMove).pairCopyOrMove
         0     0% 98.65%   253.08MB  2.20%  github.com/ncw/rclone/lib/pacer.(*Pacer).Call
         0     0% 98.65%  7231.33MB 62.79%  github.com/ncw/rclone/lib/pacer.(*Pacer).CallNoRetry
         0     0% 98.65%  7484.41MB 64.99%  github.com/ncw/rclone/lib/pacer.(*Pacer).call
         0     0% 98.65%   253.08MB  2.20%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/aws/request.(*HandlerList).Run
         0     0% 98.65%   253.08MB  2.20%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/aws/request.(*Request).Send
         0     0% 98.65%    67.01MB  0.58%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/private/protocol/rest.UnmarshalMeta
         0     0% 98.65%    67.01MB  0.58%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/private/protocol/rest.unmarshalLocationElements
         0     0% 98.65%   173.05MB  1.50%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/private/protocol/restxml.Unmarshal
         0     0% 98.65%    67.01MB  0.58%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/private/protocol/restxml.UnmarshalMeta
         0     0% 98.65%   169.54MB  1.47%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/private/protocol/xml/xmlutil.UnmarshalXML
         0     0% 98.65%    82.54MB  0.72%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/service/s3.(*S3).HeadObject
         0     0% 98.65%   169.54MB  1.47%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/service/s3.(*S3).ListObjects
         0     0% 98.65%  5581.25MB 48.46%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/service/s3/s3manager.(*multiuploader).upload
         0     0% 98.65%  7230.33MB 62.78%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/service/s3/s3manager.(*uploader).nextReader
         0     0% 98.65%  7231.33MB 62.79%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/service/s3/s3manager.(*uploader).upload
         0     0% 98.65%  7231.33MB 62.79%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/service/s3/s3manager.Uploader.Upload
         0     0% 98.65%  7231.33MB 62.79%  github.com/ncw/rclone/vendor/github.com/aws/aws-sdk-go/service/s3/s3manager.Uploader.UploadWithContext
         0     0% 98.65%   129.51MB  1.12%  net/http.(*persistConn).readResponse
         0     0% 98.65%   129.51MB  1.12%  net/http.ReadResponse
         0     0% 98.65% 10827.10MB 94.01%  sync.(*Pool).Get
```

Eventually this lead to the os killing the process in my case for running the machine OOM.

**Comment by ncw** [comment]
I'm not 100% sure this is a bug...

What it looks like to me is that 7230 MB are in use by the s3 uploader and 3596 MB are in use by rclones file buffering.

You've set transfers to 400, so that means that is 18MB/transfer in the s3 uploader and 9MB/transfer in rclone's buffers.

Rclone's buffers are set to 16MB by default (change it with the `--buffer-size` flag) and 18MB doesn't seem like an unreasonable amount of buffering for the s3 uploader - the chunks are 5MB in size and there are 2 in flight at once.

Both the internal rclone buffering and the s3uploader use `sync.Pool` which does tend to bloat memory, but not indefinitely.

Some things for you to try to use less memory
- Try `--buffer-size 0`
- Try reducing transfers - say `--transfers 200`
- Try `--s3-upload-concurrency 1`

**Comment by bkruger99** [comment]
I am a go noob (extremely),  but I'll read more on debugging and profiling to see what all can be done.

**Comment by ncw** [comment]
This was fixed in 7e6fac8b1eaeccbf48e197e2c1f12627c3e69ca5 :-)
